src/PublicApiAnalyzers/Core/Analyzers/DeclarePublicApiAnalyzer.Impl.cs

```diff
@@ -193,25 +225,36 @@ private void OnSymbolActionCore(ISymbol symbol, Action<Diagnostic> reportDiagnos
                     locationsToReport.AddRange(locations.Where(l => l.IsInSource));
                 }
 
-                var hasPublicApiEntry = _publicApiMap.TryGetValue(publicApiName, out ApiLine apiLine);
-                if (!hasPublicApiEntry)
+                ApiLine foundApiLine;
+                if (_useNullability)
                 {
-                    // Unshipped public API with no entry in public API file - report diagnostic.
-                    string errorMessageName = GetErrorMessageName(symbol, isImplicitlyDeclaredConstructor);
-                    // Compute public API names for any stale siblings to remove from unshipped text (e.g. during signature change of unshipped public API).
-                    var siblingPublicApiNamesToRemove = GetSiblingNamesToRemoveFromUnshippedText(symbol);
-                    ImmutableDictionary<string, string> propertyBag = ImmutableDictionary<string, string>.Empty
-                        .Add(PublicApiNamePropertyBagKey, publicApiName)
-                        .Add(MinimalNamePropertyBagKey, errorMessageName)
-                        .Add(PublicApiNamesOfSiblingsToRemovePropertyBagKey, siblingPublicApiNamesToRemove);
-
-                    reportDiagnosticAtLocations(DeclareNewApiRule, propertyBag, errorMessageName);
+                    var hasPublicApiEntryWithNullability = _publicApiMap.TryGetValue(publicApiName.NameWithNullability, out foundApiLine);
+                    if (!hasPublicApiEntryWithNullability)
+                    {
+                        var hasPublicApiEntryWithoutNullability = _publicApiMap.TryGetValue(publicApiName.Name, out foundApiLine);
+                        if (!hasPublicApiEntryWithoutNullability)
+                        {
+                            reportDeclareNewApi(symbol, isImplicitlyDeclaredConstructor, publicApiName.NameWithNullability);
+                        }
+                        else
+                        {
+                            reportAnnotateApi(symbol, isImplicitlyDeclaredConstructor, publicApiName, foundApiLine.IsShippedApi);
```

Comment:
I presume you have tested that FixAll works for this diagnostic? I think that would be very important functionality when enabling nullable annotations in public API files.

Reply:
It looks like we don't have test framework in place to test FixAll. I'll do some manual validation.
Please advise if there is a way to automate (I couldn't find any example in this solution nor in the roslyn-SDK that I could follow).

Reply:
Hmm I thought the below annotations did a FixAll: https://github.com/dotnet/roslyn-analyzers/blob/4578f20ea22fae2c340abca5bfad102afcf01a7d/src/PublicApiAnalyzers/UnitTests/DeclarePublicApiAnalyzerTests.cs#L980-L1001

Does it instead just fix each diagnostic separately?

Reply:
I also wonder if for discoverability of the new functionality it would be good to report a new RS diagnostic on public APIs with nullable annotations in source but nullability not enabled in public API files, say "Enable nullable annotations in public API files", that adds nullable enable to shipped/unshipped file and then adds annotations to these files. If someone doesn't care about it, they can just disable the new RS diagnostic ID for their project.

Reply:
Hum, I'm not sure how to deploy this project into a live VS instance...

Reply:
Can you try setting `Microsoft.CodeAnalysis.PublicApiAnalyzers.Setup.csproj` as the startup project and F5?

Reply:
Regarding FixAll: then my test `AnnotatedMemberInAnnotatedShippedAPI` already covers this, and I may be lacking a non-FixAll test scenario instead ;-)

Reply:
Regarding adding `#nullable enable`. I love the idea. I'll add that shortly.

Reply:
Regarding FixAll, I found out that the test framework used here automatically tests FixAll whenever testing a single Fix too. I broke the FixAll implementation and it was detected.

Reply:
Regarding manual validation: the setup project didn't seem to work as far as I could tell.

Reply:
> I found out that the test framework used here automatically tests FixAll whenever testing a single Fix too

Ah great, I forgot we did enable that auto FixAll validation in unit tests.

Reply:
> Regarding manual validation: the setup project didn't seem to work as far as I could tell.

Interesting, guess we need to investigate why it did not work. Can you file a separate tracking issue?

Reply:
Filed https://github.com/dotnet/roslyn-analyzers/issues/3153 for the setup issue.